front/src/components/WithSession.tsx

```diff
@@ -0,0 +1,60 @@
+import { ReactNode, useEffect } from "react";
+import { useDispatch } from "react-redux";
+import { useAppSelector } from "../redux/hook";
+import { setIsLoggedIn, setUser } from "../redux/slices";
+import { useLocation, useNavigate, Navigate } from "react-router-dom";
+import { checkUser } from "../api";
+
+export function WithSession({
+  children,
+}: {
+  children: ReactNode[] | ReactNode;
+}) {
+  const location = useLocation();
+  const navigate = useNavigate();
+  const { token, isLoggedIn } = useAppSelector((state) => state.session);
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    if (token) {
+      checkUser(token).then((user) => {
+        console.log("ya un user ?", user);
+        if (user) {
+          dispatch(setUser(user));
+          dispatch(setIsLoggedIn(true));
+          navigate("/");
+        } else {
+          dispatch(
+            setUser({
+              _id: "",
+              email: "",
+              password: "",
+              sub: "",
+            })
+          );
+          dispatch(setIsLoggedIn(false));
+        }
+      });
+    } else {
+      dispatch(
+        setUser({
+          _id: "",
+          email: "",
+          password: "",
+          sub: "",
+        })
+      );
+      dispatch(setIsLoggedIn(false));
+    }
+  }, [dispatch, token, navigate]);
+
+  if (
+    !isLoggedIn &&
+    location.pathname !== "/login" &&
+    location.pathname !== "/signup"
+  ) {
+    return <Navigate to="/login" replace />;
+  }
```

Comment:
Note that this will always redirect e.g. users from the `/play` page to the `/login` page because (I'm assuming?) the login credentials are not stored in a cookie. Can be done in a follow up PR, maybe just put a TODO here?

Reply:
All of this are good idea, I will try to implement them in another task